File: solutions/0001_two_sum.go
/**
 * Author: Xiangyue Cai
 * Date: 2020-03-20
 */


func twoSum(nums []int, target int) []int {
    tmp := make(map[int]int)
    for k, v := range nums {
     	if j, exists := tmp[target - v]; exists {
	    	return []int{j, k}
	 	}
	tmp[v] = k
    }
    return nil
}

File: solutions/0001_two_sum.rs
use std::collections::HashMap;

// Author: Xiangyue Cai
// Date: 2020-03-27
// O(n)
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        for (k, v) in nums.iter().enumerate() {
            let tmp = target - v;
            if map.contains_key(&tmp) {
                return vec![*map.get(&tmp).unwrap(), k as i32];
            }
            map.insert(v, k as i32);
        }
        return vec![];
    }
}